backend/rdp/peer.c

```diff
@@ -0,0 +1,352 @@
+#define _POSIX_C_SOURCE 200809L
+#include <linux/input.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/interfaces/wlr_input_device.h>
+#include <wlr/interfaces/wlr_keyboard.h>
+#include <wlr/util/log.h>
+#include "backend/rdp.h"
+#include "util/signal.h"
+
+static BOOL xf_peer_capabilities(freerdp_peer *client) {
+	return TRUE;
+}
+
+static BOOL xf_peer_post_connect(freerdp_peer *client) {
+	return TRUE;
+}
+
+static BOOL xf_peer_activate(freerdp_peer *client) {
+	struct wlr_rdp_peer_context *context =
+		(struct wlr_rdp_peer_context *)client->context;
+	struct wlr_rdp_backend *backend = context->backend;
+	rdpSettings *settings = client->settings;
+
+	if (!settings->SurfaceCommandsEnabled) {
+		wlr_log(WLR_ERROR, "RDP peer does not support SurfaceCommands");
+		return FALSE;
+	}
+
+	context->output = wlr_rdp_output_create(backend, context,
+			(int)settings->DesktopWidth, (int)settings->DesktopHeight);
+	rfx_context_reset(context->rfx_context,
+			context->output->wlr_output.width,
+			context->output->wlr_output.height);
+	nsc_context_reset(context->nsc_context,
+			context->output->wlr_output.width,
+			context->output->wlr_output.height);
+
+	if (context->flags & RDP_PEER_ACTIVATED) {
+		return TRUE;
+	}
+
+	context->pointer = wlr_rdp_pointer_create(backend);
+	// Use wlroots' software cursors instead of remote cursors
+	POINTER_SYSTEM_UPDATE pointer_system;
+	rdpPointerUpdate *pointer = client->update->pointer;
+	pointer_system.type = SYSPTR_NULL;
+	pointer->PointerSystem(client->context, &pointer_system);
+
+	context->keyboard = wlr_rdp_keyboard_create(backend, settings);
+
+	context->flags |= RDP_PEER_ACTIVATED;
+	return TRUE;
+}
+
+static int xf_suppress_output(rdpContext *context,
+		BYTE allow, const RECTANGLE_16 *area) {
+	struct wlr_rdp_peer_context *peer_context =
+		(struct wlr_rdp_peer_context *)context;
+	if (allow) {
+		peer_context->flags |= RDP_PEER_OUTPUT_ENABLED;
+	} else {
+		peer_context->flags &= (~RDP_PEER_OUTPUT_ENABLED);
+	}
+	return true;
+}
+
+static int xf_input_synchronize_event(rdpInput *input, UINT32 flags) {
+	struct wlr_rdp_peer_context *context =
+		(struct wlr_rdp_peer_context *)input->context;
+	wlr_output_damage_whole(&context->output->wlr_output);
+	return true;
+}
+
+static inline int64_t timespec_to_msec(const struct timespec *a) {
+	return (int64_t)a->tv_sec * 1000 + a->tv_nsec / 1000000;
+}
+
+static int xf_input_mouse_event(rdpInput *input,
+		UINT16 flags, UINT16 x, UINT16 y) {
+	struct wlr_rdp_peer_context *context =
+		(struct wlr_rdp_peer_context *)input->context;
+	struct wlr_input_device *wlr_device = &context->pointer->wlr_input_device;
+	struct wlr_pointer *pointer = wlr_device->pointer;
+	bool frame = false;
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+
+	if (flags & PTR_FLAGS_MOVE) {
+		struct wlr_event_pointer_motion_absolute event = { 0 };
+		int width, height;
+		wlr_output_effective_resolution(
```

Comment:
`wlr_output_effective_resolution` shouldn't be used here. Instead, `wlr_output->{width,height}` should be used and the input device should be mapped to the output.

(Applies to all absolute motion events)